operator/pkg/controllers/etcd/etcd.go

```diff
@@ -0,0 +1,346 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net"
+
+	"github.com/awslabs/kit/operator/pkg/apis/infrastructure/v1alpha1"
+	"github.com/awslabs/kit/operator/pkg/errors"
+	pkiutil "github.com/awslabs/kit/operator/pkg/utils/pki"
+
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	etcdRootCACommonName = "etcd/ca"
+)
+
+type Provider struct {
+	kubeClient client.Client
+}
+
+type CertConfig struct {
+	*certutil.Config
+	name   string
+	isCA   bool
+	caCert *x509.Certificate
+	caKey  crypto.Signer
+}
+
+func New(kubeclient client.Client) *Provider {
+	return &Provider{kubeClient: kubeclient}
+}
+
+func (e *Provider) Reconcile(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	// generate etcd service object
+	etcdServiceName, err := e.createETCDService(ctx, controlPlane)
+	if err != nil {
+		return err
+	}
+	// Create ETCD certs and keys, store them as secret in the management server
+	if err := e.createETCDCerts(ctx, controlPlane); err != nil {
+		return err
+	}
+	// TODO generate etcd stateful set
+	_ = etcdServiceName
+	return nil
+}
+
+func (e *Provider) createETCDService(ctx context.Context, controlPlane *v1alpha1.ControlPlane) (string, error) {
+	objName := etcdServiceNameFor(controlPlane.ClusterName())
+	service, err := e.getService(ctx, namespacedName(controlPlane.NamespaceName(), objName))
+	if err != nil {
+		if errors.IsNotFound(err) {
```

Comment:
controller runtime has a CreateOrUpdate utility function that's quite useful. It's effectively `apply`